cpp/oneapi/dal/algo/finiteness_checker/backend/gpu/compute_kernel_dense_dpc.cpp

```diff
@@ -0,0 +1,76 @@
+/*******************************************************************************
+* Copyright contributors to the oneDAL project
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/algo/finiteness_checker/backend/gpu/compute_kernel.hpp"
+#include "oneapi/dal/backend/primitives/reduction.hpp"
+#include "oneapi/dal/backend/primitives/utils.hpp"
+#include "oneapi/dal/detail/profiler.hpp"
+
+namespace oneapi::dal::finiteness_checker::backend {
+
+using dal::backend::context_gpu;
+using input_t = compute_input<task::compute>;
+using descriptor_t = detail::descriptor_base<task::compute>;
+
+namespace pr = dal::backend::primitives;
+
+template <typename Float>
+bool compute_finiteness(sycl::queue& queue,
+                        const pr::ndview<Float, 1>& x_1d,
+                        bool allowNaN,
+                        const dal::backend::event_vector& deps = {}) {
+    Float out;
+
+    {
+        ONEDAL_PROFILER_TASK(finiteness_checker.reduce, queue);
+        if (allowNaN) {
+            out = pr::reduce_1d(queue, x_1d, pr::logical_or<Float>{}, pr::isinf<Float>{}, deps);
+        }
+        else {
+            out =
+                pr::reduce_1d(queue, x_1d, pr::logical_or<Float>{}, pr::isinfornan<Float>{}, deps);
+        }
+    }
+    return static_cast<bool>(out);
```

Comment:
Unfortunately there is an issue with templating in reduce_1d, I'll write you about it.

Reply:
The reduce_1d defines the BinaryOp and UnaryOp via a template, and since there isn't a base functor, they cant be cast to a generalized functor to reduce_1d, this means that there are two different versions of reduce_1d being compiled, and can't be passed properly via a ternary operator (ternary assumes that the two branches yield the same type of object). Thus pre-if statement is needed.